tutorials/advanced-topics/basic-segment-tree.md

```diff
@@ -0,0 +1,179 @@
+
+For example, Analyse the image below.
+![Point Update](https://cdn.codespeedy.com/wp-content/uploads/2019/12/Node-or-Point-updates.png)
+
+Here we are update 2nd index and adding 3 to it.So all the nodes on the path back to root from it must also be updated.
+
+<Tabs>
+<TabItem value="CPP" label="CPP">
+<SolutionAuthor name="@DhruvilLakhtaria"/>
+
+```CPP
+class Solution:
+    class SegmentTree{
+
+    vector<int> v;
+    SegmentTree (vector<int>& arr, int n) {
+	    v = vector<int>(4*n);
+	    buildTree(arr, 0, n - 1, 0);
+    }          
+    
+    void buildTree(vector<int>& arr, int l, int r, int loc) {    
+	    // we have reached the leaf node
+	    if (l == r) {
+		    v[loc] = arr[l];
+		    return;
+	    } 
+	    
+	    int  mid = l + (r - l)/2;
+	    
+	    // build for left child
+	    buildTree(arr, l, mid, 2*loc + 1);
+	    // build for right child
+	    buildTree(arr, mid + 1, r, 2*loc + 2);
+ 
+	    // use above result for getting current answer.  
+	    v[loc] = v[2*loc + 1] + v[2*loc + 2];
+    } 
+      
+    int query(int loc, int l, int r, int query_l, int query_r) {
+	    // case-1: requested query interval within l-r 
+	    if (query_l <= l && query_r >= r)
+		    return v[loc];      
+	    // case-2: requested query interval is outside l-r
+	    else if (query_l > r || query_r < l)
+		    return INT_MAX;
+	    // case-3: requested query is intersecting its child.
+	    int mid = l + (r - l)/2;
+	    int left = query(2*loc + 1,l,mid,query_l, query_r);
+	    int right = query(2*loc + 2,mid + 1,r,query_l, query_r);
+	    return left + right;
+    }
+    void update(int loc, int l, int r, int update_at, int value) {
+	    if (l == r) {
+		    // update leaf node
+		    v[loc] = value;
+		    return ;
+	    }
+	    int  mid = (l + r) / 2;
+	   
+	    // if index lies in the left child call update() for left child
+	    if (update_at <= mid)
+		    update(2*loc + 1, l, mid, update_at, value);
+	    
+		// else index would lie in the right child, call update() for the right child 
+	    else
+		    update(2*loc + 2, mid + 1, r, update_at, value);
+	    
+	    // after updating the children,update the current node as smallest of its children
+	    v[loc] = v[2*loc + 1] + v[2*loc + 2];
+	 	return;  
+    }
+};
+```
+
+</TabItem>
+</Tabs>
+
+### Sample Usage
+The leetcode question demands to just fill the functions but we can also analyse our code by sample usage. 
+
+<Tabs>
+<TabItem value="CPP" label="CPP">
+<SolutionAuthor name="@DhruvilLakhtaria"/>
+
+```cpp
+vector<int> nums = {3, 1, 3, 4, 212, 4, 21};
+int n = nums.size();
+
+SegmentTree st = SegementTree(arr, n);
+// operation - 1 sum of 0-3
+st.query(0, 0, n - 1, 0, 3);
+// operation - 2 update i = 4 to 21.
+st.update(0, 0, n - 1, 4, 21);
+// operation - 3 sum of 3-6
+st.query(0, 0, n - 1, 3, 6);
+
+```
+</TabItem>
+</Tabs>
+
+
```

Comment:
Missing full solution to your example.

Reply:
The solution is complete only. Leetcode question is to only fill the functions, I have added a sample usage just incase someone wants to know how to use this segment tree class.